tools/cpp/modules_tools/generate-modmap/generate-modmap.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright 2024 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <iostream>
+#include <queue>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "generate-modmap.h"
+
+void write_modmap(std::ostream &modmap_file_stream,
+                  std::ostream &modmap_file_dot_input_stream,
+                  const std::unordered_set<ModmapItem> &modmap,
+                  const std::string &compiler) {
+  for (const auto &item : modmap) {
+    if (compiler == "clang") {
+      modmap_file_stream << "-fmodule-file=" << item.name << "=" << item.path
+                         << "\n";
+    } else if (compiler == "gcc") {
+      modmap_file_stream << item.name << " " << item.path << "\n";
```

Comment:
There should ideally be a `$root` entry in GCC modmap files, but it's up to you (the default is probably suitable).